NuRadioReco/modules/envelope_phasedarray/triggerSimulator.py

```diff
@@ -0,0 +1,220 @@
+    beam can be specified. The envelope filter is implemented using the model
+    for the ARA tunnel diode specified in utilities.diodeSimulator.
+
+    See https://arxiv.org/pdf/1903.11043.pdf
+    and https://elog.phys.hawaii.edu/elog/anita_notes/080827_041639/powertrigger.pdf
+    """
+
+    def envelope_trigger(self,
+                         station,
+                         beam_rolls,
+                         triggered_channels,
+                         threshold_factor,
+                         power_mean,
+                         power_std,
+                         output_passband=(None,200*units.MHz),
+                         cut_times=(None,None)):
+        """
+        Calculates the envelope trigger for a certain phasing configuration.
+        Beams are formed. Then, each channel to be phased is filtered with a
+        tunnel diode, the outputs are phased and, if the minimum is lower than
+        the number of antennas times (power_mean - power_std * np.abs(threshold_factor)),
+        a trigger is created.
+
+        Parameters
+        ----------
+        station: Station object
+            Description of the current station
+        beam_rolls: array of ints
+            Contains the integers for rolling the voltage traces (delays)
+        triggered_channels: array of ints
+            Ids of the triggering channels
+        threshold_factor: float
+            the threshold factor
+        power_mean: float
+            mean of the noise trace after being filtered with the diode
+        power_std: float
+            standard deviation of the noise trace after being filtered with the
+            diode. power_mean and power_std can be calculated with the function
+            calculate_noise_parameters from utilities.diodeSimulator
+        output_passband: (float, float) tuple
+            Frequencies for a 6th-order Butterworth filter to be applied after
+            the diode filtering.
+        cut_times: (float, float) tuple
+            Times for cutting the trace after diode filtering. This helps reducing
+            the number of noise-induced triggers. Doing it the other way, that is,
+            cutting and then filtering, will create two artificial pulses on the
+            edges of the trace.
+
+        Returns
+        -------
+        is_triggered: bool
+            True if the triggering condition is met
+        trigger_delays: dictionary
+            the delays for the primary channels that have caused a trigger.
+            If there is no trigger, it's an empty dictionary
+        """
+        sampling_rate = station.get_channel(0).get_sampling_rate()
+        time_step = 1. / sampling_rate
+
+        diode = diodeSimulator(output_passband)
+
+        traces = {}
+        for channel in station.iter_channels():
+            channel_id = channel.get_id()
+            if channel_id not in triggered_channels:  # skip all channels that do not participate in the trigger decision
```

Comment:
FYI: The iter_channels method has the use_channels parameter that does this automatically for you.